File: app/src/main/java/pratamawijaya/moviesinfokotlin/data/network/ApiModule.java
package pratamawijaya.moviesinfokotlin.data.network;

import android.app.Application;
import android.support.annotation.NonNull;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import dagger.Module;
import dagger.Provides;
import io.rx_cache.internal.RxCache;
import io.victoralbertos.jolyglot.GsonSpeaker;
import javax.inject.Singleton;
import okhttp3.OkHttpClient;
import pratamawijaya.moviesinfokotlin.BuildConfig;
import pratamawijaya.moviesinfokotlin.data.feature.movie.MovieCacheProviders;
import pratamawijaya.moviesinfokotlin.data.feature.movie.MovieServices;
import retrofit2.Converter;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava.RxJavaCallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;

/**
 * Created by pratama on 9/30/16.
 */
@Module @Singleton public final class ApiModule {

  @Provides @Singleton @NonNull public Retrofit provideRetrofit(@NonNull OkHttpClient okHttpClient,
      @NonNull Converter.Factory factory) {
    return new Retrofit.Builder().baseUrl(BuildConfig.SERVER_URL)
        .client(okHttpClient)
        .addConverterFactory(factory)
        .addCallAdapterFactory(RxJavaCallAdapterFactory.create())
        .validateEagerly(BuildConfig.DEBUG)
        .build();
  }

  @Provides @Singleton @NonNull Gson providesGson() {
    Gson gson = new GsonBuilder().create();
    return gson;
  }

  @Provides @Singleton @NonNull Converter.Factory provideConverter(@NonNull Gson gson) {
    return GsonConverterFactory.create(gson);
  }

  @Provides @NonNull MovieServices provideMovieServices(Retrofit retrofit) {
    return retrofit.create(MovieServices.class);
  }

  @Provides @NonNull MovieCacheProviders provideMovieCache(Application application) {
    return new RxCache.Builder().persistence(application.getCacheDir(), new GsonSpeaker())
        .using(MovieCacheProviders.class);
  }
}

File: app/src/main/java/pratamawijaya/moviesinfokotlin/data/network/ApiModule.kt
package pratamawijaya.moviesinfokotlin.data.network

import android.support.annotation.NonNull
import com.google.gson.Gson
import dagger.Module
import dagger.Provides
import okhttp3.OkHttpClient
import pratamawijaya.moviesinfokotlin.BuildConfig
import retrofit2.Converter
import retrofit2.Retrofit
import retrofit2.adapter.rxjava.RxJavaCallAdapterFactory
import retrofit2.converter.gson.GsonConverterFactory
import javax.inject.Singleton

/**
 * Created by Pratama Nur Wijaya
 * Date : Oct - 10/10/16
 * Project Name : MoviesInfoKotlin
 */

@Module
@Singleton
class ApiModule {

  @Provides
  @Singleton
  @NonNull
  fun provideRetrofit(@NonNull okHttpClient: OkHttpClient,
      @NonNull factory: Converter.Factory): Retrofit {
    return Retrofit.Builder()
        .baseUrl(BuildConfig.SERVER_URL)
        .client(okHttpClient)
        .addConverterFactory(factory)
        .addCallAdapterFactory(RxJavaCallAdapterFactory.create())
        .build()
  }

  @Provides
  @Singleton
  @NonNull
  fun providesGson(): Gson {
    return Gson()
  }

  @Provides
  @Singleton
  @NonNull
  fun provideConverter(@NonNull gson: Gson): Converter.Factory {
    return GsonConverterFactory.create(gson)
  }

}